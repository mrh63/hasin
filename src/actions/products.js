import { actionTypes } from './types';
import axios from 'axios';


//REMOVE_PRODUCT
export const removeProduct = ({id} = {}) => ({
    type: actionTypes.REMOVE_PRODUCT,
    id
});

//SET_PRODUCTS
export const setProducts = (products) => ({
    type: actionTypes.SET_PRODUCTS,
    products
});

export const startSetProducts = (removeItem = []) => async dispatch => {

    const products = [];

    const res = await axios.get(['http://127.0.0.1:3000', 'data'].join('/'))
    .catch(function(thrown) {
        console.log('server not available');
        return dispatch(setProducts(products));
    });

    for (let key in res.data) {
        if (removeItem.indexOf(key) === -1)
            products.push({productKey: key, ...res.data[key]});
    }
    dispatch(setProducts(products));
}