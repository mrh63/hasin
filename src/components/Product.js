import React from 'react';
import { useSelector } from 'react-redux';
import { Container, Row, Col } from 'reactstrap';

const Product = (props) => {

    const products = useSelector(state => state.products);
    const product = products.find(product => product.title === props.match.params.title);
    if (!product) {

    }
    const {
        title = 'product not found',
        image = '/images/no-image.jpg',
        description = 'product not found'
    } = product || {};
    
    return (
        <div className="product">
            <Container>
                <Row className="d-flex justify-content-center">
                    <Col lg="8">
                        <img width="100%" src={image} alt={title} />
                    </Col>
                </Row>
                <Row className="mt-5 d-flex justify-content-center">
                    <Col lg="8">
                        <p>{description}</p>
                    </Col>
                </Row>
            </Container>
        </div>
    )
}

export default Product;