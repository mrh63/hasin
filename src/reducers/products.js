import { actionTypes } from '../actions/types';

//Products Reducer
const productsReducerDefaultState = [];

export default (state = productsReducerDefaultState, action) => {
    switch (action.type) {
        case actionTypes.REMOVE_PRODUCT:
            return state.filter(({id}) => id !== action.id);
        case actionTypes.SET_PRODUCTS:
            return action.products;
        default:
            return state;
    }
};