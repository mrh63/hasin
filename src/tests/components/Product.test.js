import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Provider} from 'react-redux'
import { mount } from 'enzyme';
import Product from '../../components/Product';
import products from '../fixtures/products';

const createMockStore = configureMockStore([thunk]);

let store, match, getWrapper;

beforeEach(() => {
    store = createMockStore({
        products
    });
    getWrapper = (match) => mount(
        <Provider store={store}>
            <Product match={match} />
        </Provider>
    )
})

test('should render Product correctly with valid route', () => {
    match = {params: {title: "مایکت"}};
    const wrapper = getWrapper(match);
    expect(wrapper).toMatchSnapshot();
});

test('should render Product correctly with invalid route', () => {
    match = {params: {title: "abc"}};
    const wrapper = getWrapper(match);
    expect(wrapper).toMatchSnapshot();
});



