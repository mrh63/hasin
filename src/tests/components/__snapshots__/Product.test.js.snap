// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Product correctly with invalid route 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Product
    match={
      Object {
        "params": Object {
          "title": "abc",
        },
      }
    }
  >
    <div
      className="product"
    >
      <Container
        tag="div"
      >
        <div
          className="container"
        >
          <Row
            className="d-flex justify-content-center"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <div
              className="d-flex justify-content-center row"
            >
              <Col
                lg="8"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="col-lg-8"
                >
                  <img
                    alt="product not found"
                    src="/images/no-image.jpg"
                    width="100%"
                  />
                </div>
              </Col>
            </div>
          </Row>
          <Row
            className="mt-5 d-flex justify-content-center"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <div
              className="mt-5 d-flex justify-content-center row"
            >
              <Col
                lg="8"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="col-lg-8"
                >
                  <p>
                    product not found
                  </p>
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </div>
  </Product>
</Provider>
`;

exports[`should render Product correctly with valid route 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Product
    match={
      Object {
        "params": Object {
          "title": "مایکت",
        },
      }
    }
  >
    <div
      className="product"
    >
      <Container
        tag="div"
      >
        <div
          className="container"
        >
          <Row
            className="d-flex justify-content-center"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <div
              className="d-flex justify-content-center row"
            >
              <Col
                lg="8"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="col-lg-8"
                >
                  <img
                    alt="مایکت"
                    src="/images/myket.jpg"
                    width="100%"
                  />
                </div>
              </Col>
            </div>
          </Row>
          <Row
            className="mt-5 d-flex justify-content-center"
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <div
              className="mt-5 d-flex justify-content-center row"
            >
              <Col
                lg="8"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="col-lg-8"
                >
                  <p>
                    مایکت یک استور اندرویدی با بیش از ۳۰۰ هزار برنامه و بازی ا‌ست که از پاییز ۱۳۹۰ فعالیت خود را آغاز کرده. در حال حاضر بیش از ۳۳ هزار توسعه‌دهنده ایرانی و خارجی با مایکت همکاری می‌کنند و برنامه‌ها و بازی‌های تولید شده آن‌ها، پس از بررسی در دسترس کاربران قرار می‌گیرد.         در حال حاضر مایکت با بیش از 12 میلیون کاربر توانسته جایگاه خود را به عنوان یکی از پرطرفدارترین مارکت‌های ایرانی ثابت کند؛ مایکت روزانه آپدیت می‌شود و به صورت آنلاین پاسخ‌گوی سوالات و مشکلات احتمالی کاربران و توسعه‌دهندگان است. اگر برنامه یا بازی مد نظر شما در مایکت نبود، آن را سفارش دهید. در کمترین زمان ممکن تیم حرفه ای و درجه یک مایکت فراهمش می‌کنند!         برای دانلود مایکت و استفاده از آن، می‌توانید به سایت آن مراجعه نمایید.
                  </p>
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </div>
  </Product>
</Provider>
`;
